### PR TITLE
Add time option to traffic_ctl host down.  

### PR DESCRIPTION
**Pull request by jrushford**
Add time option to traffic_ctl host down.  The --time option allows the user to mark a host down for a
specific amount of time in seconds after which the host will be marked back up.  A time of '0', the default, marks the host down indefinitely and requires manual restoration using 'traffic_ctl host up'.

### CODE REVIEW

**Comment by zwoop** [comment]
[approve ci clang-format]

File: mgmt/api/CoreAPI.cc [hunk 1]
```diff
@@ -885,10 +885,11 @@ EventSignalCbUnregister(const char *event_name, TSEventSignalFunc func)
  *-------------------------------------------------------------------------
  * Sets the HOST status to Down
  */
+
 TSMgmtError
-HostStatusSetDown(const char *name)
+HostStatusSetDown(const char *req, int len)
```
Comment (zwoop) [inline mgmt/api/CoreAPI.cc]: Is "len" here the actual downtime? If so, "len" seems very confusing, looking at this API, I'd assume len here is the length of the const char* req. if it is so, then this is fine, if not, we should change this to down_time or some such.

File: mgmt/api/CoreAPI.h [hunk 2]
```diff
@@ -82,6 +82,6 @@ TSMgmtError EventIsActive(const char *event_name, bool *is_current);
 TSMgmtError EventSignalCbRegister(const char *event_name, TSEventSignalFunc func, void *data);
 TSMgmtError EventSignalCbUnregister(const char *event_name, TSEventSignalFunc func);
 
-TSMgmtError HostStatusSetDown(const char *name);
+TSMgmtError HostStatusSetDown(const char *name, const int down_time);
```
Comment (zwoop) [inline mgmt/api/CoreAPI.h]: I never do "const" on an int like this, which is call by value. It's harmless, but can be confusing.

File: proxy/HostStatus.cc [hunk 3]
```diff
@@ -85,24 +107,48 @@ HostStatus::setHostStatus(const char *name, HostStatus_t status)
...
+  HostStatRec_t *host_stat = nullptr;
   ink_rwlock_wrlock(&host_status_rwlock);
-  ink_hash_table_insert(hosts_statuses, name, reinterpret_cast<void *>(status));
+  if (ink_hash_table_lookup(hosts_statuses, name, reinterpret_cast<InkHashTableValue *>(&host_stat)) == 0) {
+    host_stat = static_cast<HostStatRec_t *>(ats_malloc(sizeof(HostStatRec_t)));
+    ink_hash_table_insert(hosts_statuses, name, reinterpret_cast<InkHashTableValue *>(host_stat));
+  }
+  host_stat->status                                                 = status;
+  host_stat->down_time                                              = down_time;
+  status == HostStatus_t::HOST_STATUS_DOWN ? host_stat->marked_down = time(0) : host_stat->marked_down = 0;
```
Comment (zwoop) [inline proxy/HostStatus.cc]: This reads really poorly, do we really have to use a tertiary here? There's a lot of assignments in there ... And clang-format loses it ...

**Comment by mlibbey** [comment]
- I don't really get the use case for this. Usually when our folks take a machine out of production for some sort of maintenance, they want to make sure the maintenance is really done before putting it back into rotation -- I'd think they'd be scared this option would automatically put it back in before everything was done. (So, I'm guessing there is a different use case for this?)
- Does this persist across trafficserver restarts and machine reboots? The documentation should indicate the behavior.

**Comment by jrushford** [comment]
@zwoop requested changes have been made.

@mlibbey 

- I think that is most valuable to temporarily remove a parent from use due to load or when troubleshooting rather than removing said parent from config files and reloading configs. 
  
- @vmamidi and I discussed persisting this but decided not to do so now.  That would be done at a later time once we switch to using  @a-canary host status data.

File: mgmt/LocalManager.h [hunk 4]
```diff
@@ -94,7 +94,7 @@ class LocalManager : public BaseManager
   void processDrain(int to_drain = 1);
   void rollLogFiles();
   void clearStats(const char *name = nullptr);
-  void hostStatusSetDown(const char *name);
+  void hostStatusSetDown(const char *req, int len);
```
Comment (zwoop) [inline mgmt/LocalManager.h]: It's still len here ?

File: mgmt/LocalManager.cc [hunk 5]
```diff
@@ -122,9 +122,9 @@ LocalManager::rollLogFiles()
 }
 
 void
-LocalManager::hostStatusSetDown(const char *name)
+LocalManager::hostStatusSetDown(const char *req, int len)
```
Comment (zwoop) [inline mgmt/LocalManager.cc]: Still len?

File: mgmt/api/include/mgmtapi.h [hunk 6]
```diff
@@ -488,7 +488,7 @@ tsapi TSMgmtError TSReadFromUrl(char *url, char **header, int *headerSize, char
  */
 tsapi TSMgmtError TSReadFromUrlEx(const char *url, char **header, int *headerSize, char **body, int *bodySize, int timeout);
 tsapi TSMgmtError TSHostStatusSetUp(const char *name);
-tsapi TSMgmtError TSHostStatusSetDown(const char *name);
+tsapi TSMgmtError TSHostStatusSetDown(const int down_time, const char *name);
```
Comment (zwoop) [inline mgmt/api/include/mgmtapi.h]: I think this should be TSHostStatusSetDown(const char*name, int down_time);   I also don't think it has to be "const int" on down_time.

**Comment by zwoop** [comment]
Most important here is the TSHostStatusSetDown(const char*name, int down_time);  change, but please look at the "len" naming too.

**Comment by jrushford** [comment]
[approve ci]

**Comment by jrushford** [comment]
[approve ci autest]

**Comment by jrushford** [comment]
@zwoop in CoreAPIRemote.cc, HostStatusSetDown(), the host_name and down_time are serialized and remain serialized until un-marshaling occurs when HostStatus::mgmt_host_status_down_callback() is called.  So the arguments 'req' and 'len' are correct.  I fixed up the arguments and hope it's clearer now.

**Comment by vmamidi** [comment]
@zwoop can you look at this again? @jrushford has done the requested changes.

File: mgmt/LocalManager.cc (hunk 5 above)
Comment (vmamidi) [reply mgmt/LocalManager.cc]: Should arguments be renamed to indicate it is marshaled data?

File: mgmt/LocalManager.h (hunk 4 above)
Comment (vmamidi) [reply mgmt/LocalManager.h]: Should arguments be renamed to indicate it is marshaled data?

**Pull request merged**